File: Greg/python/complex.py
class Complex:
  def __init__(self, real, img):
    self.real = real
    self.img = img
  def label(self):
    return "(" + str(self.real) + ". " + str(self.img) + ")"
  def add(self,x):
    r = self.real + x.real
    i = self.img + x.img
    return Complex(r, i)
  def times(self,y):
    r = self.real*y.real - self.img*y.img
    i = self.img*y.real + self.real*y.img
    return Complex(r, i)

if __name__ == "__main__":
  c = Complex(1, 2)
  d = Complex(3, - 4)
  
  # should be (1 + 2i)
  print(c.label())
  e = c.add(d)
  # should be (4 - 2i)
  print(e.label())
  # should be (1 + 2i)
  print(c.label())
  # should be (3 - 4i)
  print(d.label())

  # should be (11 + 2i)
  f = c.times(d)
  print(f.label())

  # what should this be?
  d = d.times(d)
  print(d.label())

File: rex/cpp/complex.cc
#include<iostream>
using namespace std;

class Complex{
  public:
    Complex(double a,double b);
    Complex add(Complex c);
    Complex sub(Complex c);
    Complex times(Complex c);
    void label();
  private:
    double _a;
    double _b;
};
Complex::Complex(double a ,double b):_a(a),_b(b){}

Complex Complex::add(Complex C){
  double c = C._a;
  double d = C._b;
  return Complex(_a+c,_b+d);
}
Complex Complex::sub(Complex C){
  double c = C._a;
  double d = C._b;
  return Complex(_a-c,_b-d);
}
Complex Complex::times(Complex C){
  double c = C._a;
  double d = C._b;
  return Complex(_a*c-_b*d,_a*d+_b*c);
  
}
void Complex::label(){
  cout<<"("<<_a<<"+"<<_b<<"i)"<<endl;
}

int main() {
  Complex c(1, 2);
  Complex d(3, -4);

  c.label(); // (1 + 2i)
  
  Complex e = c.add(d);
  e.label(); // (4 - 2i)
  c.label(); // (1 + 2i)
  d.label(); // (3 - 4i)
  
  Complex f = c.times(d);
  f.label(); // (11 + 2i)

  d = d.times(d);
  d.label(); // what should this be?
}